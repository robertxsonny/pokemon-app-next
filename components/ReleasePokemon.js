import { useEffect, useState } from 'react';
import Link from 'next/link';
import styled from '@emotion/styled';

import Modal from './Modal';
import { SentenceCase, SmallCtaButton, TextButton } from '../styles/shared';

import { useMyPokemons } from '../hooks/my-pokemon';

const DangerTextButton = styled(TextButton)({
  color: 'red',
  marginTop: 12
})

const DangerLinkButton = DangerTextButton.withComponent('a');

const DangerCtaButton = styled(SmallCtaButton)({
  backgroundColor: 'red',
  '&:hover:not([disabled])': {
    backgroundColor: 'firebrick'
  },
  '&:active:not([disabled])': {
    backgroundColor: 'darkred'
  }
})

const ReleasePokemon = ({ open, nickname, onClose }) => {
  const { releasePokemon } = useMyPokemons();
  const [released, setReleased] = useState(false);

  useEffect(() => {
    if (open) {
      setReleased(false)
    }
  }, [open])

  const onReleasePokemon = () => {
    releasePokemon(nickname);
    setReleased(true);
  }

  const nameWihSentenceCase = <SentenceCase>{nickname}</SentenceCase>

  if (released) {
    return (
      <Modal open={open} title={<>{nameWihSentenceCase} has been released</>} emoji="🕊️">
        <Link href="/my-pokemon">
          <DangerLinkButton>Go back to my collection</DangerLinkButton>
        </Link>
      </Modal>
    )
  }

  return (
    <Modal open={open} title={<>Are you sure you want to release {nameWihSentenceCase}?</>} emoji="⚠️">
      <DangerCtaButton onClick={onReleasePokemon}>Yes, release from collection</DangerCtaButton>
      <DangerTextButton onClick={onClose}>No, cancel</DangerTextButton>
    </Modal>
  )
}

export default ReleasePokemon;
