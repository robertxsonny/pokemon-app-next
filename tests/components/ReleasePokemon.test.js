import { render, screen, fireEvent } from '@testing-library/react';
import ReleasePokemon from '../../components/ReleasePokemon';

jest.mock('../../hooks/my-pokemon', () => ({
  useMyPokemons: () => ({
    releasePokemon: jest.fn()
  })
}))

describe('ReleasePokemon component', () => {
  test('Should render confirmation screen at first', () => {
    const { container } = render(<ReleasePokemon open nickname="pikachu" />);
    expect(container).toMatchSnapshot();
  })

  test('Should render notification after confirming to release Pokemon', () => {
    const { container } = render(<ReleasePokemon open nickname="pikachu" />);
    fireEvent.click(screen.getByRole('button', { name: 'Yes, release from collection' }));
    expect(container).toMatchSnapshot();
  })

  test('Should not render anything if not opened', () => {
    const { container } = render(<ReleasePokemon open={false}/>);
    expect(container).toMatchSnapshot();
  })
})