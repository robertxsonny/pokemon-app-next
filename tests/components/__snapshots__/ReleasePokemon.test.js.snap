// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReleasePokemon component Should not render anything if not opened 1`] = `<div />`;

exports[`ReleasePokemon component Should render confirmation screen at first 1`] = `
<div>
  <div
    class="css-1f6dfq1-ModalContainer e2hvhql2"
  >
    <div
      class="css-pujlrx-ModalContent e2hvhql1"
    >
      <span
        aria-label="[object Object]"
        class="css-1k4msm7-Emoji e55skjj7"
        role="img"
      >
        ‚ö†Ô∏è
      </span>
      <h3
        class="css-1f0is5a-ModalTitle e2hvhql0"
      >
        Are you sure you want to release 
        <span
          class="css-8aj3s4-SentenceCase e55skjj8"
        >
          pikachu
        </span>
        ?
      </h3>
      <button
        class="css-11qewao-CtaButton-SmallCtaButton-DangerCtaButton ejb9ufw0"
      >
        Yes, release from collection
      </button>
      <button
        class="css-hqh9ye-TextButton-DangerTextButton ejb9ufw1"
      >
        No, cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`ReleasePokemon component Should render notification after confirming to release Pokemon 1`] = `
<div>
  <div
    class="css-1f6dfq1-ModalContainer e2hvhql2"
  >
    <div
      class="css-pujlrx-ModalContent e2hvhql1"
    >
      <span
        aria-label="[object Object]"
        class="css-1k4msm7-Emoji e55skjj7"
        role="img"
      >
        üïäÔ∏è
      </span>
      <h3
        class="css-1f0is5a-ModalTitle e2hvhql0"
      >
        <span
          class="css-8aj3s4-SentenceCase e55skjj8"
        >
          pikachu
        </span>
         has been released
      </h3>
      <a
        class="css-1jmjyid-DangerLinkButton-TextButton-DangerTextButton ejb9ufw2"
        href="/my-pokemon"
      >
        Go back to my collection
      </a>
    </div>
  </div>
</div>
`;
